CollaborationService/js/CollaborationService.js

```diff
@@ -0,0 +1,1004 @@
+        firebaseKey = collaborativeCollection.firebaseKey,
+        addFirebaseKeyToValues, sendRequestWithParentIfPresent;
+
+    addFirebaseKeyToValues = function (snapshot) {
+      var values = snapshot.val() || {};
+      values.values = values.values || {};
+      values.values[firebaseKey] = snapshot.key;
+      return values;
+    };
+
+    sendRequestWithParentIfPresent = function (request, values) {
+      var parent = values.parent || {},
+          parentCollectionName = parent[FIREBASE_PARENT_COLLECTION_NAME_KEY],
+          parentCollaborationKeyName = parentCollectionName ? self.collaborationContexts[contextName].collections[parentCollectionName].firebaseKey : null,
+          parentCollaborationKeyValue = parent[FIREBASE_PARENT_COLLABORATION_KEY_VALUE_KEY];
+
+      if (parentCollectionName && parentCollaborationKeyName && parentCollaborationKeyValue) {
+        self.findCaseByCollaborationKey(contextName, parentCollectionName, parentCollaborationKeyName, parentCollaborationKeyValue, function (_case) {
+          if (_case) {
+            request.values.parent = _case.id;
+          }
+          dispatcher.sendRequest(request);
+        });
+      }
+      else {
+        dispatcher.sendRequest(request);
+      }
+    };
+
+    casesRef.on('child_added', function (snapshot) {
+      self.findCaseByCollaborationKey(contextName, collectionName, firebaseKey, snapshot.key, function (_case) {
+        if (!_case) {
+          var values = addFirebaseKeyToValues(snapshot);
+          sendRequestWithParentIfPresent({
+            action: 'create',
+            resource: 'dataContext[' + contextName + '].collection[' + collectionName + '].case',
+            values: {
+              values: values.values
+            }
+          }, values);
+        }
+      });
+    });
+
+    casesRef.on('child_changed', function (snapshot) {
+      self.findCaseByCollaborationKey(contextName, collectionName, firebaseKey, snapshot.key, function (_case) {
+        if (_case) {
+          var values = addFirebaseKeyToValues(snapshot);
+          sendRequestWithParentIfPresent({
+            action: 'update',
+            resource: 'dataContext[' + contextName + '].collection[' + collectionName + '].caseByID[' + _case.id + ']',
+            values: {
+              values: values.values
+            }
+          }, values);
+        }
+      });
+    });
+
+    casesRef.on('child_removed', function (snapshot) {
+      self.findCaseByCollaborationKey(contextName, collectionName, firebaseKey, snapshot.key, function (_case) {
+        if (_case) {
+          sendRequestWithParentIfPresent({
+            action: 'delete',
+            resource: 'dataContext[' + contextName + '].collection[' + collectionName + '].caseByID[' + _case.id + ']'
+          }, snapshot.val());
+        }
+      });
+    });
+  }
+
+}).init();
+
+var dispatcher = Object.create({ // jshint ignore:line
+
+    init: function () {
+      firebase.initializeApp({
```

Comment:
There is no payment it's just being able to login to administer the Firebase project  I just added you (jsandoe@concord.org) as another "owner" of the project which means you have full permissions to do anything.  You should get an invitation from Firebase.